README.md
# Fluent plugin for Amazon Kinesis

[![Gitter](https://badges.gitter.im/awslabs/aws-fluent-plugin-kinesis.svg)](https://gitter.im/awslabs/aws-fluent-plugin-kinesis?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
[![Build Status](https://travis-ci.org/awslabs/aws-fluent-plugin-kinesis.svg?branch=master)](https://travis-ci.org/awslabs/aws-fluent-plugin-kinesis)
[![Gem Version](https://badge.fury.io/rb/fluent-plugin-kinesis.svg)](https://rubygems.org/gems/fluent-plugin-kinesis-modified)

[Fluentd][fluentd] output plugin
that sends events to [Amazon Kinesis Data Streams][streams] and [Amazon Kinesis Data Firehose][firehose]. Also it supports [KPL Aggregated record format][kpl]. This gem includes four output plugins respectively:

- `kinesis_streams`
- `kinesis_firehose`
- `kinesis_streams_aggregated`
- `kinesis_streams_aggregated_modified`

Also, there is a [documentation on Fluentd official site][fluentd-doc-kinesis].

**Note**: This README is for v3. Plugin v3 is almost compatible with v2. If you use v1, see the [old README][v1-readme].

## Installation
This Fluentd plugin is available as the `fluent-plugin-kinesis` gem from RubyGems.

    gem install fluent-plugin-kinesis-modified

Or you can install this plugin for [td-agent][td-agent] as:

    td-agent-gem install fluent-plugin-kinesis-modified

If you would like to build by yourself and install, please see the section below. Your need [bundler][bundler] for this.

In case of using with Fluentd: Fluentd will be also installed via the process below.

    git clone -b feature_modified_aggregation --single-branch https://github.com/BlackNight95/aws-fluent-plugin-kinesis.git
    cd aws-fluent-plugin-kinesis
    bundle install
    bundle exec rake build
    bundle exec rake install

Also, you can use this plugin with td-agent: You have to install td-agent before installing this plugin.

    git clone -b feature_modified_aggregation --single-branch https://github.com/BlackNight95/aws-fluent-plugin-kinesis.git
    cd aws-fluent-plugin-kinesis
    bundle install
    bundle exec rake build
    fluent-gem install pkg/fluent-plugin-kinesis-modified

Or just download specify your Ruby library path. Below is the sample for specifying your library path via RUBYLIB.

    git clone -b feature_modified_aggregation --single-branch https://github.com/BlackNight95/aws-fluent-plugin-kinesis.git
    cd aws-fluent-plugin-kinesis
    bundle install
    export RUBYLIB=$RUBYLIB:/path/to/aws-fluent-plugin-kinesis/lib

## Dependencies
 * Ruby 2.1.0+
 * Fluentd 0.14.10+

## Basic Usage
Here are general procedures for using this plugin:

 1. Install.
 1. Edit configuration
 1. Run Fluentd or td-agent

You can run this plugin with Fluentd as follows:

 1. Install.
 1. Edit configuration file and save it as 'fluentd.conf'.
 1. Then, run `fluentd -c /path/to/fluentd.conf`

To run with td-agent, it would be as follows:

 1. Install.
 1. Edit configuration file provided by td-agent.
 1. Then, run or restart td-agent.

## Getting started
Assume you use Amazon EC2 instances with Instance profile. If you want to use specific credentials, see [Credentials](#configuration-credentials).

### kinesis_streams
    <match your_tag>
      @type kinesis_streams
      region us-east-1
      stream_name your_stream
      partition_key key  # Otherwise, use random partition key
    </match>
For more details, see [Configuration: kinesis_streams](#configuration-kinesis_streams).

### kinesis_firehose
    <match your_tag>
      @type kinesis_firehose
      region us-east-1
      delivery_stream_name your_stream
    </match>
For more details, see [Configuration: kinesis_firehose](#configuration-kinesis_firehose).

### kinesis_streams_aggregated/kinesis_streams_aggregated_modified
    <match your_tag>
      @type kinesis_streams_aggregated/kinesis_streams_aggregated_modified
      region us-east-1
      stream_name your_stream
      # Unlike kinesis_streams, there is no way to use dynamic partition key.
      # fixed_partition_key or random.
    </match>
For more details, see [Configuration: kinesis_streams_aggregated](#configuration-kinesis_streams_aggregated) and [Configuration: kinesis_streams_aggregated_modified](#configuration-kinesis_streams_aggregated_modified).

### For better throughput
Add configurations like below:

      flush_interval 1
      chunk_limit_size 1m
      flush_thread_interval 0.1
      flush_thread_burst_interval 0.01
      flush_thread_count 15

When you use Fluent v1.0 (td-agent3), write these configurations in buffer section. For more details, see [Buffer section configurations](https://docs.fluentd.org/articles/buffer-section).

Note: Each value should be adjusted to your system by yourself.

## Configuration: Credentials
To put records into Amazon Kinesis Data Streams or Firehose, you need to provide AWS security credentials somehow. Without specifying credentials in config file, this plugin automatically fetch credential just following AWS SDK for Ruby does (environment variable, shared profile, and instance profile).

This plugin uses the same configuration in [fluent-plugin-s3][fluent-plugin-s3].

**aws_key_id**

AWS access key id. This parameter is required when your agent is not running on EC2 instance with an IAM Role. When using an IAM role, make sure to configure `instance_profile_credentials`. Usage can be found below.

**aws_sec_key**

AWS secret key. This parameter is required when your agent is not running on EC2 instance with an IAM Role.

**aws_iam_retries**

The number of attempts to make (with exponential backoff) when loading instance profile credentials from the EC2 metadata service using an IAM role. Defaults to 5 retries.

### assume_role_credentials
Typically, you can use AssumeRole for cross-account access or federation.

    <match *>
      @type kinesis_streams

      <assume_role_credentials>
        role_arn          ROLE_ARN
        role_session_name ROLE_SESSION_NAME
      </assume_role_credentials>
    </match>

See also:

*   [Using IAM Roles - AWS Identity and Access
    Management](http://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use.html)
*   [Aws::STS::Client](http://docs.aws.amazon.com/sdkforruby/api/Aws/STS/Client.html)
*   [Aws::AssumeRoleCredentials](http://docs.aws.amazon.com/sdkforruby/api/Aws/AssumeRoleCredentials.html)

**role_arn (required)**

The Amazon Resource Name (ARN) of the role to assume.

**role_session_name (required)**

An identifier for the assumed role session.

**policy**

An IAM policy in JSON format.

**duration_seconds**

The duration, in seconds, of the role session. The value can range from 900 seconds (15 minutes) to 3600 seconds (1 hour). By default, the value is set to 3600 seconds.

**external_id**

A unique identifier that is used by third parties when assuming roles in their customers' accounts.

**sts_http_proxy**

Proxy url for proxying requests to amazon sts service api. This needs to be  set up independently from global http_proxy parameter for the use case in which requests to kinesis api are going via kinesis vpc endpoint but requests to sts api have to go via http proxy.
It should be added to assume_role_credentials configuration stanza in the next format:
    sts_http_proxy http://[username:password]@hostname:port

### instance_profile_credentials

Retrieve temporary security credentials via HTTP request. This is useful on EC2 instance.

    <match *>
      @type kinesis_streams

      <instance_profile_credentials>
        ip_address IP_ADDRESS
        port       PORT
      </instance_profile_credentials>
    </match>

See also:

*   [Aws::InstanceProfileCredentials](http://docs.aws.amazon.com/sdkforruby/api/Aws/InstanceProfileCredentials.html)
*   [Temporary Security Credentials - AWS Identity and Access
    Management](http://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp.html)
*   [Instance Metadata and User Data - Amazon Elastic Compute
    Cloud](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html)

**retries**

Number of times to retry when retrieving credentials. Default is 5.

**ip_address**

Default is 169.254.169.254.

**port**

Default is 80.

**http_open_timeout**

Default is 5.

**http_read_timeout**

Default is 5.

### shared_credentials

This loads AWS access credentials from local ini file. This is useful for local developing.

    <match *>
      @type kinesis_streams

      <shared_credentials>
        path         PATH
        profile_name PROFILE_NAME
      </shared_credentials>
    </match>

See also:

*   [Aws::SharedCredentials](http://docs.aws.amazon.com/sdkforruby/api/Aws/SharedCredentials.html)

**path**

Path to the shared file. Defaults to "#{Dir.home}/.aws/credentials".

**profile_name**

Defaults to 'default' or `[ENV]('AWS_PROFILE')`.

### process_credentials

This loads AWS access credentials from an external process.

    <match *>
      @type kinesis_streams

      <process_credentials>
        process CMD
      </process_credentials>
    </match>

See also:

*   [Aws::ProcessCredentials](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/ProcessCredentials.html)
*   [Sourcing Credentials From External Processes](https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#sourcing-credentials-from-external-processes)

**process (required)**

Command to be executed as an external process.

## Configuration: Format

### format (section)
This plugin uses `Fluent::TextFormatter` to serialize record to string. See [formatter.rb] for more details. By default, it uses `json` formatter same as specific like below:

    <match *>
      @type kinesis_streams

      <format>
        @type json
      </format>
    </match>

For other configurations of `json` formatter, see [json Formatter Plugin](https://docs.fluentd.org/articles/formatter_json).

### inject (section)
This plugin uses `Fluent::TimeFormatter` and other injection configurations. See [inject.rb] for more details.

For example, the config below will add `time` field whose value is event time with nanosecond and `tag` field whose value is its tag.

    <match *>
      @type kinesis_streams

      <inject>
        time_key time
        tag_key tag
      </inject>
    </match>

By default, `time_type string` and `time_format %Y-%m-%dT%H:%M:%S.%N%z` are already set to be applicable to Elasticsearch sub-second format. Although, you can use any configuration.

Also, there are some format related options below:

### data_key
If your record contains a field whose string should be sent to Amazon Kinesis directly (without formatter), use this parameter to specify the field. In that case, other fields than **data_key** are thrown away and never sent to Amazon Kinesis. Default `nil`, which means whole record will be formatted and sent.

### compression
Specifing compression way for data of each record. Current accepted options are `zlib`. Otherwise, no compression will be preformed.

### log_truncate_max_size
Integer, default 1024. When emitting the log entry, the message will be truncated by this size to avoid infinite loop when the log is also sent to Kinesis. The value 0 means no truncation.

### chomp_record
Boolean. Default `false`. If it is enabled, the plugin calls chomp and removes separator from the end of each record. This option is for compatible format with plugin v2. See [#142](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/142) for more details.  
When you use [kinesis_firehose](#kinesis_firehose) output, [append_new_line](#append_new_line) option is `true` as default. If [append_new_line](#append_new_line) is enabled, the plugin calls chomp as [chomp_record](#chomp_record) is `true` before appending `\n` to each record. Therefore, you don't need to enable [chomp_record](#chomp_record) option when you use [kinesis_firehose](#kinesis_firehose) with default configuration. If you want to set [append_new_line](#append_new_line) `false`, you can choose [chomp_record](#chomp_record) `false` (default) or `true` (compatible format with plugin v2).

## Configuration: API
### region
AWS region of your stream. It should be in form like `us-east-1`, `us-west-2`. Refer to [Regions and Endpoints in AWS General Reference][region] for supported regions.

Default `nil`, which means try to find from environment variable `AWS_REGION`.

### max_record_size
The upper limit of size of each record. Default is 1 MB which is the limitation of Kinesis.

### http_proxy
HTTP proxy for API calling. Default `nil`.

### endpoint
API endpoint URL, for testing. Default `nil`.

### ssl_verify_peer
Boolean. Disable if you want to verify ssl connection, for testing. Default `true`.

### debug
Boolean. Enable if you need to debug Amazon Kinesis Data Firehose API call. Default is `false`.

## Configuration: Batch request
### retries_on_batch_request
Integer, default is 8. The plugin will put multiple records to Amazon Kinesis Data Streams in batches using PutRecords. A set of records in a batch may fail for reasons documented in the Kinesis Service API Reference for PutRecords. Failed records will be retried **retries_on_batch_request** times. If a record fails all retries an error log will be emitted.

### reset_backoff_if_success
Boolean, default `true`. If enabled, when after retrying, the next retrying checks the number of succeeded records on the former batch request and reset exponential backoff if there is any success. Because batch request could be composed by requests across shards, simple exponential backoff for the batch request wouldn't work some cases.

### batch_request_max_count
Integer, default 500. The number of max count of making batch request from record chunk. It can't exceed the default value because it's API limit.

Default:

- `kinesis_streams`: 500
- `kinesis_firehose`: 500
- `kinesis_streams_aggregated`: 100,000
- `kinesis_streams_aggregated_modified`: 100,000

### size_kb_per_record
Integer. The number of max size of making batch request from record chunk. It can't exceed the default value because it's API limit.

Default:

- `kinesis_streams`: 5 MB
- `kinesis_firehose`: 4 MB
- `kinesis_streams_aggregated`: 1 MB
- `kinesis_streams_aggregated_modified`: 1 MB

## Configuration: kinesis_streams
Here are `kinesis_streams` specific configurations.

### stream_name
Name of the stream to put data.

### partition_key
A key to extract partition key from JSON object. Default `nil`, which means partition key will be generated randomly.

## Configuration: kinesis_firehose
Here are `kinesis_firehose` specific configurations.

### delivery_stream_name
Name of the delivery stream to put data.

### append_new_line
Boolean. Default `true`. If it is enabled, the plugin adds new line character (`\n`) to each serialized record.  
Before appending `\n`, plugin calls chomp and removes separator from the end of each record as [chomp_record](#chomp_record) is `true`. Therefore, you don't need to enable [chomp_record](#chomp_record) option when you use [kinesis_firehose](#kinesis_firehose) output with default configuration ([append_new_line](#append_new_line) is `true`). If you want to set [append_new_line](#append_new_line) `false`, you can choose [chomp_record](#chomp_record) `false` (default) or `true` (compatible format with plugin v2).

## Configuration: kinesis_streams_aggregated
Here are `kinesis_streams_aggregated` specific configurations.

### stream_name
Name of the stream to put data.

### fixed_partition_key
A value of fixed partition key. Default `nil`, which means partition key will be generated randomly.

Note: if you specified this option, all records go to a single shard.

## Configuration: kinesis_streams_aggregated_modified
Here are `kinesis_streams_aggregated_modified` specific configurations.

### stream_name
Name of the stream to put data.

### fixed_partition_key
A value of fixed partition key. Default `nil`, which means partition key will be generated randomly.

Note: if you specified this option, all records go to a single shard.

### max_records_per_call
Integer. Max number of aggregated records sent to kinesis with put_records method in each iteration. max_records_per_call * size_kb_per_record should not exceed 5 MB.


## Development

To launch `fluentd` process with this plugin for development, follow the steps below:

    git clone https://github.com/awslabs/aws-fluent-plugin-kinesis.git
    cd aws-fluent-plugin-kinesis
    make # will install gems dependency
    bundle exec fluentd -c /path/to/fluent.conf

To launch using specified version of Fluentd, use `BUNDLE_GEMFILE` environment variable:

    BUNDLE_GEMFILE=$PWD/gemfiles/Gemfile.td-agent-3.3.0 bundle exec fluentd -c /path/to/fluent.conf

## Contributing

Bug reports and pull requests are welcome on [GitHub][github].

## Related Resources

* [Amazon Kinesis Data Streams Developer Guide](http://docs.aws.amazon.com/kinesis/latest/dev/introduction.html)
* [Amazon Kinesis Data Firehose Developer Guide](http://docs.aws.amazon.com/firehose/latest/dev/what-is-this-service.html)

[fluentd]: http://fluentd.org/
[streams]: https://aws.amazon.com/kinesis/streams/
[firehose]: https://aws.amazon.com/kinesis/firehose/
[kpl]: https://github.com/awslabs/amazon-kinesis-producer/blob/master/aggregation-format.md
[td-agent]: https://github.com/treasure-data/td-agent
[bundler]: http://bundler.io/
[region]: http://docs.aws.amazon.com/general/latest/gr/rande.html#ak_region
[fluentd_buffer]: http://docs.fluentd.org/articles/buffer-plugin-overview
[github]: https://github.com/awslabs/aws-fluent-plugin-kinesis
[formatter.rb]: https://github.com/fluent/fluentd/blob/master/lib/fluent/formatter.rb
[inject.rb]: https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin_helper/inject.rb
[fluentd-doc-kinesis]: http://docs.fluentd.org/articles/kinesis-stream
[fluent-plugin-s3]: https://github.com/fluent/fluent-plugin-s3
[v1-readme]: https://github.com/awslabs/aws-fluent-plugin-kinesis/blob/v1/README.md


CHANGELOG.md
# CHANGELOG

## 3.1.2

- Enhancement - Reserve `kinesis_streams_aggregated` from 3.1.0
- Enhancement - Change modified `kinesis_streams_aggregated` name to `kinesis_streams_aggregated_modified`

## 3.1.1

- Enhancement - Modify `kinesis_streams_aggregated` to improve throughput efficiency
- Enhancement - Change parameter name batch_request_max_size to size_kb_per_record
- Enhancement - Add max_records_per_call parameter
- Bug - Fix empty batch error in split_to_batches in api.rb

## 3.1.0

- Feature - Add process_credentials configuration : [#178](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/178)

## 3.0.0

Plugin v3 is almost compatible with v2. You can use v3 with the same configuration as v2. For more details, please see [README](README.md).

- Enhancement - Use modularized AWS SDK v3 since fluent-plugin-s3 also supports it : [#152](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/152)
- Enhancement - Remove support for Fluentd v0.12 and use new Plugin API : [#156](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/156) (also fix [#133](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/133))
- Enhancement / Breaking change - Remove *chomp* method from internal formatter : [#142](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/142) [#144](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/144)
- Enhancement - Add *chomp_record* option for compatible format with plugin v2 : [#142](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/142)
- Bug - Fix undefined method error in flushing buffers : [#133](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/133)
- Bug - Fix dependency problem on AWS SDK : [#161](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/161)

## 2.1.1

- Bug - Fix require aws-sdk-core before requiring the aws related libraries [#140](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/140)

## 2.1.0

- Feature - Added sts_http_proxy parameter to assume_role_credentials configuration [#136](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/136)

## 2.0.1

- Bug - Fix AWS SDK conflict with s3 plugin [#131](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/131)

## 2.0.0

- Feature - Add `kinesis_streams_aggregated` ouput plugin [#107](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/107)
- Feature - Support fluentd worker model [#104](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/104)
- Feature - Support AWS SDK for Ruby v3 [#102](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/102)
- Enhancement - Refactor class design [#103](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/103)
- Enhancement - More configuration for AssumeRole [#63](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/63)
- Enhancement - Refactor credentials helper [#94](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/94)
- Enhancement - Revisit backoff logic [#69](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/69)
- Enhancement - Support compressing output [#98](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/98)
- Enhancement - Support nanosecond time_key format [#124](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/124)
- License - Move back to Apache License Version 2.0

## 1.3.0

- Feature - Log KPL stdout/err to logger [#129](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/129) [#130](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/130)

## 1.2.0

- Feature - Add reduce_max_size_error_message configuration [#127](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/127)
- Bug fix - Fix empty batch error [#125](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/125)

## 1.1.3

- Bug fix - Fix issues with fluentd 0.14.12 [#99](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/99)

## 1.1.2

- Bug fix - Adjust credentials_provider for newer version of aws-ruby-sdk [#93](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/93)

## 1.1.1

- Bug fix - Fix incompatibility for AWS SDK 2.5 [#80](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/80)
- Bug fix - Fix wrong logic for batch request [#81](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/81)

## 1.1.0

- Feature - Derive stream name from fluentd tag for KPL [#67](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/67)
- Enhancement - Make http_proxy parameter secret [#64](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/64)
- Bug fix - Plugin incompatible with new fluentd release 0.14 [#70](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/70)
- Misc - Fix legacy test and reduce travis tests [#74](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/74)
- Misc - Some test, benchmark improvement [#74](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/74), [#75](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/75), [#76](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/76)

## 1.0.1

- Bug fix - Instance profile credentials expiring [#58](https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/58)

## 1.0.0

To support Firehose and KPL, this was refactored and added more tests.

- Feature - Support Firehose
- Feature - Support KPL
- Feature - New parameters, such as formatter, data_key, reset_backoff_if_success, batch_request_max_count, batch_request_max_size, log_truncate_max_size

## 0.4.0

- Feature - Add option to ensure Kinesis Stream connection. [#35](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/35)
- Feature - Add option to support zlib compression. [#39](https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/39)

Note: We introduced [Semantic Versioning](http://semver.org/) here.

## 0.3.6


lib/fluent/plugin/out_kinesis_streams_aggregated.rb
#
# Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.

require 'fluent/plugin/kinesis'
require 'fluent/plugin/kinesis_helper/aggregator'

module Fluent
  module Plugin
    class KinesisStreamsAggregatedOutput < KinesisOutput
      Fluent::Plugin.register_output('kinesis_streams_aggregated', self)
      include KinesisHelper::Aggregator::Mixin

      RequestType = :streams_aggregated
      BatchRequestLimitCount = 100_000
      BatchRequestLimitSize  = 1024 * 1024
      include KinesisHelper::API::BatchRequest

      config_param :stream_name, :string
      config_param :fixed_partition_key, :string, default: nil

      def configure(conf)
        super
        @partition_key_generator = create_partition_key_generator
        @size_kb_per_record -= offset
        @max_record_size -= offset
      end

      def format(tag, time, record)
        format_for_api do
          [@data_formatter.call(tag, time, record)]
        end
      end

      # def write(chunk)
      #   write_records_batch(chunk) do |batch|
      #     key = @partition_key_generator.call
      #     records = batch.map{|(data)|data}
      #     client.put_records(
      #       stream_name: @stream_name,
      #       records: [{
      #         partition_key: key,
      #         data: aggregator.aggregate(records, key),
      #       }],
      #     )
      #   end
      # end

      def write(chunk)
        write_records_batch2(chunk) do |batches|
          records_to_send = []
          batches.each do |batch|
            key = @partition_key_generator.call
            records = batch.map{|(data)|data}
            regularized_agg_record = {
              partition_key: key,
              data: aggregator.aggregate(records, key)
            }
            records_to_send << regularized_agg_record
          end
          client.put_records(
            stream_name: @stream_name,
            records: records_to_send
          )
        end
      end

      def offset
        @offset ||= AggregateOffset + @partition_key_generator.call.size*2
      end

      private

      def size_of_values(record)
        super(record) + RecordOffset
      end

      def create_partition_key_generator
        if @fixed_partition_key.nil?
          ->() { SecureRandom.hex(16) }
        else
          ->() { @fixed_partition_key }
        end
      end
    end
  end
end


lib/fluent/plugin/out_kinesis_streams_aggregated_modified.rb
#
# Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.

require 'fluent/plugin/kinesis'
require 'fluent/plugin/kinesis_helper/aggregator'

module Fluent
  module Plugin
    class KinesisStreamsAggregatedOutput < KinesisOutput
      Fluent::Plugin.register_output('kinesis_streams_aggregated_modified', self)
      include KinesisHelper::Aggregator::Mixin

      RequestType = :streams_aggregated
      BatchRequestLimitCount = 100_000
      BatchRequestLimitSize  = 1024 * 1024
      include KinesisHelper::API::BatchRequest

      config_param :stream_name, :string
      config_param :fixed_partition_key, :string, default: nil

      def configure(conf)
        super
        @partition_key_generator = create_partition_key_generator
        @size_kb_per_record -= offset
        @max_record_size -= offset
      end

      def format(tag, time, record)
        format_for_api do
          [@data_formatter.call(tag, time, record)]
        end
      end

      def write(chunk)
        write_records_batch2(chunk) do |batches|
          records_to_send = []
          batches.each do |batch|
            key = @partition_key_generator.call
            records = batch.map{|(data)|data}
            regularized_agg_record = {
              partition_key: key,
              data: aggregator.aggregate(records, key)
            }
            records_to_send << regularized_agg_record
          end
          client.put_records(
            stream_name: @stream_name,
            records: records_to_send
          )
        end
      end

      def offset
        @offset ||= AggregateOffset + @partition_key_generator.call.size*2
      end

      private

      def size_of_values(record)
        super(record) + RecordOffset
      end

      def create_partition_key_generator
        if @fixed_partition_key.nil?
          ->() { SecureRandom.hex(16) }
        else
          ->() { @fixed_partition_key }
        end
      end
    end
  end
end


lib/fluent/plugin/kinesis.rb
#
# Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.

require 'fluent/plugin/output'
require 'fluent/plugin/kinesis_helper/client'
require 'fluent/plugin/kinesis_helper/api'
require 'zlib'

module Fluent
  module Plugin
    class KinesisOutput < Fluent::Plugin::Output
      include Fluent::MessagePackFactory::Mixin
      include KinesisHelper::Client
      include KinesisHelper::API

      class SkipRecordError < ::StandardError
        def initialize(message, record)
          super message
          @record_message = if record.is_a? Array
                              record.reverse.map(&:to_s).join(', ')
                            else
                              record.to_s
                            end
        end

        def to_s
          super + ": " + @record_message
        end
      end
      class KeyNotFoundError < SkipRecordError
        def initialize(key, record)
          super "Key '#{key}' doesn't exist", record
        end
      end
      class ExceedMaxRecordSizeError < SkipRecordError
        def initialize(size, record)
          super "Record size limit exceeded in #{size/1024} KB", record
        end
      end
      class InvalidRecordError < SkipRecordError
        def initialize(record)
          super "Invalid type of record", record
        end
      end

      config_param :data_key,              :string,  default: nil
      config_param :log_truncate_max_size, :integer, default: 1024
      config_param :compression,           :string,  default: nil

      desc "Formatter calls chomp and removes separator from the end of each record. This option is for compatible format with plugin v2. (default: false)"
      # https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/142
      config_param :chomp_record,          :bool,    default: false

      config_section :format do
        config_set_default :@type, 'json'
      end
      config_section :inject do
        config_set_default :time_type, 'string'
        config_set_default :time_format, '%Y-%m-%dT%H:%M:%S.%N%z'
      end

      config_param :debug, :bool, default: false
      config_param :max_records_per_call, :integer, default: 128

      helpers :formatter, :inject

      def configure(conf)
        super
        @data_formatter = data_formatter_create(conf)
      end

      def multi_workers_ready?
        true
      end

      private

      def data_formatter_create(conf)
        formatter = formatter_create
        compressor = compressor_create
        if @data_key.nil?
          if @chomp_record
            ->(tag, time, record) {
              record = inject_values_to_record(tag, time, record)
              # Formatter calls chomp and removes separator from the end of each record.
              # This option is for compatible format with plugin v2.
              # https://github.com/awslabs/aws-fluent-plugin-kinesis/issues/142
              compressor.call(formatter.format(tag, time, record).chomp.b)
            }
          else
            ->(tag, time, record) {
              record = inject_values_to_record(tag, time, record)
              compressor.call(formatter.format(tag, time, record).b)
            }
          end
        else
          ->(tag, time, record) {
            raise InvalidRecordError, record unless record.is_a? Hash
            raise KeyNotFoundError.new(@data_key, record) if record[@data_key].nil?
            compressor.call(record[@data_key].to_s.b)
          }
        end
      end

      def compressor_create
        case @compression
        when "zlib"
          ->(data) { Zlib::Deflate.deflate(data) }
        else
          ->(data) { data }
        end
      end

      def format_for_api(&block)
        converted = block.call
        size = size_of_values(converted)
        if size > @max_record_size
          raise ExceedMaxRecordSizeError.new(size, converted)
        end
        converted.to_msgpack
      rescue SkipRecordError => e
        log.error(truncate e)
        ''
      end

      def write_records_batch(chunk, &block)
        unique_id = chunk.dump_unique_id_hex(chunk.unique_id)
        chunk.open do |io|
          records = msgpack_unpacker(io).to_enum
          split_to_batches(records) do |batch, size|
            log.debug(sprintf "Write chunk %s / %3d records / %4d KB", unique_id, batch.size, size/1024)
            batch_request_with_retry(batch, &block)
            log.debug("Finish writing chunk")
          end
        end
      end

      def write_records_batch2(chunk, &block)
        unique_id = chunk.dump_unique_id_hex(chunk.unique_id)
        chunk.open do |io|
          records = msgpack_unpacker(io).to_enum
          batches = []
          sizes = []
          split_to_batches(records) do |batch, size|
            batches << batch
            sizes << size            
          end
          (0..batches.size).step(@max_records_per_call) do |i|
            batches_slice = batches[i..i+@max_records_per_call-1]
            size = sizes[i..i+@max_records_per_call-1].inject(:+)
            batches_size = batches_slice.map(&:size).inject(:+)
            if not size.nil?
              log.debug(sprintf "Write chunk %s / %3d batches / %3d records / %4d KB", unique_id, batches_slice.size, batches_size, size/1024)
              batch_request_with_retry(batches_slice, &block)
              log.debug("Finish writing chunk")
            end
          end
        end
      end

      def request_type
        self.class::RequestType
      end

      def truncate(msg)
        if @log_truncate_max_size == 0 or (msg.to_s.size <= @log_truncate_max_size)
          msg.to_s
        else
          msg.to_s[0...@log_truncate_max_size]
        end
      end
    end
  end
end
